perun-contract/src/Perun/Offchain.hs

```diff
@@ -0,0 +1,572 @@
+      let AssetClass (s, tn) = ct
+      if valueOf (_ciTxOutValue o) s tn == 1
+        then -- Note: If any check fails at this point, we can safely throw an error
+        -- because we know that the utxo at this address carries the ThreadToken.AbortCase
+        -- Therefore, no other utxo can carry the ThreadToken, as it is an NFT.
+        case _ciTxOutScriptDatum o of
+          (_, Just (Datum e)) -> case PlutusTx.fromBuiltinData e of
+            Nothing -> throwing _FindChannelError WrongDatumTypeError
+            Just d@ChannelDatum {..} -> if (channelTokenAsset channelToken) == ct && cID == getChannelId channelParameters && (channelTokenName (channelHash cID)) == tn then return (oref, o, d) else throwing _FindChannelError InvalidDatumError
+          _otherwise -> throwing _FindChannelError DatumMissingError
+        else filterChannelUtxos xs
+
+addFunding :: Integer -> Integer -> [Integer] -> [Integer]
+addFunding amount idx f =
+  let (s, e) = genericSplitAt idx f
+   in s ++ amount : P.drop 1 e
+
+type TokenState = Last ChannelToken
+
+tellToken :: (AsPerunError e, AsContractError e) => ChannelToken -> Contract TokenState s e ()
+tellToken = tell . Last . Just
+
+--
+-- Top-level contract, exposing all endpoints.
+--
+contract :: Contract TokenState ChannelSchema PerunError ()
+contract =
+  selectList
+    [ start',
+      fund',
+      abort',
+      open',
+      dispute',
+      close',
+      forceClose'
+    ]
+    >> contract
+  where
+    start' = endpoint @"start" start
+    fund' = endpoint @"fund" fund
+    abort' = endpoint @"abort" abort
+    open' = endpoint @"open" open
+    dispute' = endpoint @"dispute" dispute
+    close' = endpoint @"close" close
+    forceClose' = endpoint @"forceClose" forceClose
+
+compressSignatures :: AllSignedStates -> SignedState
+compressSignatures (AllSignedStates state sigs) = SignedState sigs (Scripts.datumHash (Datum (PlutusTx.toBuiltinData state))) state
+
+makeAllSignedStates :: [(SignedMessage ChannelState)] -> AllSignedStates
+makeAllSignedStates sms =
+  let datum = getDatum . osmDatum $ head sms
+      state = fromJust $ PlutusTx.fromBuiltinData datum
+   in AllSignedStates state (map osmSignature sms)
+
+-- | isValidChannelStart checks if transaction represented by the given list of inputs,
+-- and the given *single* channel output is a valid opening transaction for a channel with
+-- the given channel id.
+isValidChannelStart :: [TxOutRef] -> (Address, Value, ChannelDatum) -> ChannelID -> Bool
+isValidChannelStart inputs (outAddress, outValue, ChannelDatum {..}) cId =
+  let ref = ctTxOutRef channelToken
+      s = channelTokenSymbol ref
+      tn = channelTokenName (channelHash cId)
+   in ref `elem` inputs
+        && getChannelId channelParameters == cId
+        && outAddress == channelAddress cId
+        && valueOf outValue s tn == 1
+        && channelToken == ChannelToken s tn ref
+        && not disputed
+        && False -- TODO: Complete these checks!
```

Comment:
This is never used and was merely intended as a sketch on what must hold for something to be considered a valid channel start. I'm confident that we resemble this where we actually verify the validity of channel starts.

```suggestion
-- | isValidChannelStart checks if transaction represented by the given list of inputs,
-- and the given *single* channel output is a valid opening transaction for a channel with
-- the given channel id.
isValidChannelStart :: [TxOutRef] -> (Address, Value, ChannelDatum) -> ChannelID -> Bool
isValidChannelStart inputs (outAddress, outValue, ChannelDatum {..}) cId =
  let ref = ctTxOutRef channelToken
      s = channelTokenSymbol ref
      tn = channelTokenName (channelHash cId)
   in ref `elem` inputs
        && getChannelId channelParameters == cId
        && outAddress == channelAddress cId
        && valueOf outValue s tn == 1
        && channelToken == ChannelToken s tn ref
        && not disputed
        && False -- TODO: Complete these checks!
```

Reply:
True, removed.